hal/include/HAL/Encoder.h

```diff
@@ -9,26 +9,57 @@
-                        int32_t* status);  // TODO: fix routing
-void freeEncoder(void* encoder_pointer, int32_t* status);
-void resetEncoder(void* encoder_pointer, int32_t* status);
-int32_t getEncoder(void* encoder_pointer, int32_t* status);  // Raw value
-double getEncoderPeriod(void* encoder_pointer, int32_t* status);
-void setEncoderMaxPeriod(void* encoder_pointer, double maxPeriod,
-                         int32_t* status);
-bool getEncoderStopped(void* encoder_pointer, int32_t* status);
-bool getEncoderDirection(void* encoder_pointer, int32_t* status);
-void setEncoderReverseDirection(void* encoder_pointer, bool reverseDirection,
+enum EncoderIndexingType {
+  kResetWhileHigh,
+  kResetWhileLow,
+  kResetOnFallingEdge,
+  kResetOnRisingEdge
+};
+enum EncoderEncodingType { k1X, k2X, k4X };
+
+HalEncoderHandle initializeEncoder(
+    uint8_t port_a_module, uint32_t port_a_pin, bool port_a_analog_trigger,
+    uint8_t port_b_module, uint32_t port_b_pin, bool port_b_analog_trigger,
+    bool reverseDirection, EncoderEncodingType encodingType, int32_t* status);
+void freeEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+int32_t getEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+int32_t getRawEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+int32_t getEncodingScaleEncoder(HalEncoderHandle encoder_handle,
                                 int32_t* status);
-void setEncoderSamplesToAverage(void* encoder_pointer,
-                                uint32_t samplesToAverage, int32_t* status);
-uint32_t getEncoderSamplesToAverage(void* encoder_pointer, int32_t* status);
-void setEncoderIndexSource(void* encoder_pointer, uint32_t pin,
-                           bool analogTrigger, bool activeHigh,
-                           bool edgeSensitive, int32_t* status);
+void resetEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+int32_t getPeriodEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+void setMaxPeriodEncoder(HalEncoderHandle encoder_handle, double maxPeriod,
+                         int32_t* status);
+uint8_t getStoppedEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+uint8_t getDirectionEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+double getDistanceEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+double getRateEncoder(HalEncoderHandle encoder_handle, int32_t* status);
+void setMinRateEncoder(HalEncoderHandle encoder_handle, double minRate,
+                       int32_t* status);
+void setDistancePerPulseEncoder(HalEncoderHandle encoder_handle,
+                                double distancePerPulse, int32_t* status);
+void setReverseDirectionEncoder(HalEncoderHandle encoder_handle,
+                                uint8_t reverseDirection, int32_t* status);
+void setSamplesToAverageEncoder(HalEncoderHandle encoder_handle,
+                                int32_t samplesToAverage, int32_t* status);
```

Comment:
What's the reason for the naming style change to put Encoder at the end rather than the beginning?  This is inconsistent with the JNI naming and the other functions (e.g. getCounterX)